src/libraries/System.Private.CoreLib/src/System/ComAwareWeakReference.cs

```diff
@@ -0,0 +1,179 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+#if FEATURE_COMINTEROP || FEATURE_COMWRAPPERS
+
+using static System.WeakReferenceHandleTags;
+
+namespace System
+{
+    internal sealed partial class ComAwareWeakReference
+    {
+        private readonly nint _weakHandle;
+        private ComInfo? _comInfo;
+
+        internal sealed class ComInfo
+        {
+            private readonly IntPtr _pComWeakRef;
+            private readonly long _wrapperId;
+
+            internal object? ResolveTarget()
+            {
+                return ComWeakRefToObject(_pComWeakRef, _wrapperId);
+            }
+
+            // see: syncblk.h
+            private const int IS_HASHCODE_BIT_NUMBER = 26;
+            private const int BIT_SBLK_IS_HASHCODE = 1 << IS_HASHCODE_BIT_NUMBER;
+            private const int BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX = 0x08000000;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal static unsafe ComInfo? FromObject(object? target)
+            {
+                if (target == null)
+                    return null;
+
+                fixed (byte* pRawData = &target.GetRawData())
+                {
+                    // The header is 4 bytes before MT field on all architectures
+                    int header = *(int*)(pRawData - sizeof(IntPtr) - sizeof(int));
+                    // common case: target does not have a syncblock, so there is no interop info
+                    if ((header & (BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX | BIT_SBLK_IS_HASHCODE)) != BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX)
+                        return null;
+                }
+
+                return FromObjectSlow(target);
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            private static unsafe ComInfo? FromObjectSlow(object target)
+            {
+                IntPtr pComWeakRef = ObjectToComWeakRef(target, out long wrapperId);
+                if (pComWeakRef == 0)
+                    return null;
+
+                return new ComInfo(pComWeakRef, wrapperId);
+            }
+
+            private ComInfo(IntPtr pComWeakRef, long wrapperId)
+            {
+                Debug.Assert(pComWeakRef != IntPtr.Zero);
+                _pComWeakRef = pComWeakRef;
+                _wrapperId = wrapperId;
+            }
+
+            ~ComInfo()
+            {
+                Marshal.Release(_pComWeakRef);
+            }
+        }
+
+        private ComAwareWeakReference(nint weakHandle)
+        {
+            Debug.Assert(weakHandle != 0);
+            _weakHandle = weakHandle;
+        }
+
+        ~ComAwareWeakReference()
+        {
+            GCHandle.InternalFree(_weakHandle);
```

Comment:
The class is intentionally internal - to make sure we control use of its instances. We should never expose it publicly and each instance should have only one reference to it - from the corresponding `WeakReference` via a strong handle.

That should guarantee:
- noone can do `lock(this)` and therefore we can use `lock(this)` ourselves.
- finalization is truly final. Since the owning handle is severed when WeakReference is finalized, the ComAwareWeakReference instance is completely unreachable from other objects when it is finalized.

I think we do not need to worry about resurrection, finalization while still constructing, use after finalization, finalization concurrent with use and other finalization issues.  
These assumptions would not work with publicly exposed instances, but I think it is safe to assume here.

Reply:
Right, it should all work fine when there are no bugs. Setting the `_weakHandle` to null is a very cheap way to ensure that we get a nice crash in case there are bugs.